Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/threshold/BasicThresholder.java

```diff
@@ -206,95 +162,147 @@ public double getPrimaryThreshold() {
+     * automatically scaled to be calibrated with the average score (bounded below
+     * by an absolute constant like 0.7)
+     * 
+     * @param factor    the factor being scaled
+     * @param method    transformation method
+     * @param dimension the dimension of the problem (currently unused)
+     * @return a scaled value of the factor
+     */
+
+    protected double adjustedFactor(double factor, TransformMethod method, int dimension) {
+        double correctedFactor = factor;
+        double base = primaryDeviation.getMean();
+        if (autoThreshold && base < lowerThreshold && method != TransformMethod.NORMALIZE) {
+            correctedFactor = primaryDeviation.getMean() * factor / lowerThreshold;
         }
+        return max(correctedFactor, MINIMUM_Z_FACTOR);
+    }
 
-        if (intermediateFraction == 1 && threshold > lowerThreshold) {
-            // calculating grade based on factor
-            double longTermDeviation = longTermDeviation();
-            double tFactor = upperZfactor;
-            if (longTermDeviation > 0) {
-                tFactor = Math.min(tFactor, (score - primaryDeviation.getMean()) / longTermDeviation);
-            }
-            return (tFactor - zFactor) / (upperZfactor - zFactor);
+    /**
+     * The following computes the standard deviation of the scores. But we have
+     * multiple ways of measuring that -- if the scores are typically symmetric then
+     * many of these measures concide. However transformation of the values may
+     * cause the score distribution to be unusual. For example, if NORMALIZATION is
+     * used then the scores (below the average) end up being close to the average
+     * (an example of the asymmetry) and thus only standard deviation is used. But
+     * for other distributions we could directly estimate the deviation of the
+     * scores below the dynamic mean in an online manner, and we do so in
+     * thresholdDeviation. An orthogonal component is the effect of
+     * shingling/differencing which connect up the scores from consecutive input.
+     * 
+     * @param method      transformation method
+     * @param shingleSize shinglesize used
+     * @return an estimate of long term deviation from mean of a stochastic series
+     */
+    protected double longTermDeviation(TransformMethod method, int shingleSize) {
+
+        if (shingleSize == 1
+                && !(method == TransformMethod.DIFFERENCE || method == TransformMethod.NORMALIZE_DIFFERENCE)) {
+            // control the effect of large values above a threshold from raising the
+            // threshold
+            return min(sqrt(2.0) * thresholdDeviation.getDeviation(), primaryDeviation.getDeviation());
         } else {
-            double upper = max(upperThreshold, 2 * threshold);
-            double quasiScore = Math.min(score, upper);
-            return (quasiScore - threshold) / (upper - threshold);
+            double first = primaryDeviation.getDeviation();
```

Comment:
It is possible and likely that the standard deviation is small. It is hard to say "not representative" though -- because we are unsure about what we are trying to represent (anomalies). The hope is that we can use the literature on variable normalization to tune specific settings down the road. So it's not about ok/not-ok. This is what the best we can do at this point of time without further investigation.

Reply:
I still don't follow. I meant to ask why not use scores below the dynamic mean as in other cases and why small deviation is useful here. I can ask you offline.

Reply:
Ah - the other measures such as thresholdDeviation is even smaller :) Basically Normalization is changing the scores in a very interesting way such that primaryDeviation is much larger than thresholdDeviation !!